### PR TITLE
HYDRA-1226 - Support translation of openPBR

### PR DESCRIPTION
**Pull request by lilike-adsk**
Support translation of openPBR

Note: The emission parts needs to be revised when maya's change is ready.

### CODE REVIEW

File: lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp [hunk 1]
```diff
@@ -395,6 +395,38 @@ class MayaHydraUvAttrConverter : public MayaHydraMaterialAttrConverter
...
+        if (emissionColor.IsHolding<GfVec3f>()) {
+            emissionColorVec3f = emissionColor.UncheckedGet<GfVec3f>();
+        }
+
+        // TODO: Use emissionWeight directly when it's available in Maya
+        float emissionWeightFloat = 0.0f;
+        VtValue emissionLuminance = MayaHydraMaterialNetworkConverter::ConvertMayaAttrToValue(
+            node, "emissionLuminance", SdfValueTypeNames->Float, fallback, outPlug);
+        if (emissionLuminance.IsHolding<float>()) {
+            emissionWeightFloat = emissionLuminance.UncheckedGet<float>() / 1000.f;
```
Comment (lanierd-adsk) [inline lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp]: Could you please add a comment as why you are dividing by 1000 ?

File: lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp [hunk 2]
```diff
@@ -484,6 +516,54 @@ class MayaHydraTransmissionMaterialAttrConverter : public MayaHydraComputedMater
...
+            // effect in hydra.
+            val = 1.0e-4f;
+        }
+
+        float fGeometryOpacity = 1.0f;
+        if (geometryOpacity.IsHolding<float>()) {
+            fGeometryOpacity = geometryOpacity.UncheckedGet<float>();
+        }
+
+        val *= fGeometryOpacity;
```
Comment (lanierd-adsk) [inline lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp]: I think you could move this to line after 558 in the if block.

File: lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp [hunk 3]
```diff
@@ -657,6 +762,21 @@ void MayaHydraMaterialNetworkConverter::initialize()
...
+            {
+                { MayaHydraAdapterTokens->diffuseColor, openPBRBaseColorConverter },
+                { MayaHydraAdapterTokens->emissiveColor, openPBREmissionColorConverter },
+                { MayaHydraAdapterTokens->specularColor, openPBRSpecularColorConverter },
+                { MayaHydraAdapterTokens->ior, openPBRSpecularIORConverter },
+                { MayaHydraAdapterTokens->roughness, openPBRSpecularRoughnessConverter },
+                { MayaHydraAdapterTokens->clearcoat, openPBRCoatConverter },
+                { MayaHydraAdapterTokens->clearcoatRoughness, openPBRCoatRoughnessConverter },
+                { MayaHydraAdapterTokens->opacity, openPBRTransmissionToOpacity },
+                { MayaHydraAdapterTokens->metallic, openPBRMetallicConverter },
```
Comment (lanierd-adsk) [inline lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp:775]: I think you should set the useSpecularWorkflow as well :
{ MayaHydraAdapterTokens->useSpecularWorkflow, fixedOneInt }, so the specular color is used :

useSpecularWorkflow - int - 0 (by default)

This node can fundamentally operate in two modes : Specular workflow where you provide a texture/value to the specularColor input. Or, Metallic workflow where you provide a texture/value to the metallic input. Depending on the 0 or 1 value of this parameter, the following parameters are conditionally enabled:

File: test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/emissionColor.png
Comment (lanierd-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/emissionColor.png:1]: emmisionColor.jpg and emissionLuminance.jpg look the same to me ? Could we differentiate the 2 images with different colors/settings ? Samr for geometryOpacity.jpg.

File: test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularRoughness.png
Comment (lanierd-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularRoughness.png:1]: I would put some crazy value in the specular roughness as for me this image is the same as the previous specularIOR.png.

File: test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularWeight.png
Comment (lanierd-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularWeight.png:1]: I don't see any difference with the previous image, could we push it to the limits to see some changes ?

**Review by lanierd-adsk — COMMENTED** [review]
Thanks for doing this translation. Only a fews minor comments.

**Comment by lanierd-adsk** [comment]
FYI : @ppt-adsk @debloip-adsk

File: lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp (hunk 1 above)
Comment (lilike-adsk) [reply lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp]: This is a temp simulation currently used. also applied in MaterialX's translation, will add a comment

File: lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp (hunk 2 above)
Comment (lilike-adsk) [reply lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp]: yes, we can.

File: lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp (hunk 3 above)
Comment (lilike-adsk) [reply lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp:775]: hmm, why we didn't do that for standardSurface here? I remember we handled that logic somewhere else?

Comment (lilike-adsk) [reply lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp:775]: I saw it, it was already handled here for all material converters

        // If we are using a specular color which is not white, the UsdPreviewsurface specular
        // workflow must be enabled to use the specular color which is done by setting the
        // UsdPreviewSurface param "useSpecularWork" to 1
        {
            const auto it = material.parameters.find(_specularColorToken);
            if (it != material.parameters.cend()) {
                const VtValue& specColorVal = it->second;
                if (!specColorVal.IsEmpty()
                    && specColorVal.UncheckedGet<GfVec3f>() != GfVec3f(1, 1, 1)) {
                    material.parameters[_useSpecularWorkflowToken] = VtValue(1);
                }
            }
        }

File: lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp (hunk 2 above)
Comment (lanierd-adsk) [reply lib/mayaHydra/hydraExtensions/adapters/materialNetworkConverter.cpp]: Could you please combine line 560 with line 557 ?

File: test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularRoughness.png
Comment (lanierd-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularRoughness.png:1]: This image is pure black ? I don't think that's on purpose, right ?

File: test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularIOR.png
Comment (lanierd-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularIOR.png:1]: This image is pure black ? I don't think that's on purpose, right ?

File: test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularWeight.png
Comment (lanierd-adsk) [inline test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularWeight.png:1]: This image is pure black ? Is that on purpose ?

File: test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularRoughness.png
Comment (lilike-adsk) [reply test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularRoughness.png:1]: No, it's not black, there're differences if you download look them locally

Comment (lanierd-adsk) [reply test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularRoughness.png:1]: Honestly, I would do more contrasted images, to be sure that our image comparison system catch it when the image is not what it should be.

Comment (lilike-adsk) [reply test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularRoughness.png:1]: Already tried, unfortunately the specular part is special on this test scene, there's no big gap from eyes if you set a different specularWeight, and even worse for IOR as it's for refraction

Comment (lilike-adsk) [reply test/lib/mayaUsd/render/mayaToHydra/OpenPBRSurfaceTest/specularRoughness.png:1]: The good thing is the ImageDiff tool can catch the diff/error if the specularWeight doesn't work.

**Review by lanierd-adsk — APPROVED** [review]

**Pull request merged**